src/universal/components/ErrorBoundary.js

```diff
@@ -0,0 +1,36 @@
+import React, {Component} from 'react';
+import ErrorComponent from 'universal/components/ErrorComponent/ErrorComponent';
+import PropTypes from 'prop-types';
+
+class ErrorBoundary extends Component {
+  static propTypes = {
+    children: PropTypes.element.isRequired
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = { error: null, errorInfo: null };
+  }
+
+  componentDidCatch(error, errorInfo) {
+    // Catch errors in any components below and re-render with error message
+    this.setState({
+      error,
+      errorInfo
+    });
+    // You can also log error messages to an error reporting service here
+  }
+
+  render() {
+    if (this.state.errorInfo) {
+      // Error path
+      return (
+        <ErrorComponent error={{message: this.state.error}} />
+      );
+    }
+    // Normally, just render children
+    return this.props.children;
```

Comment:
that's right, i just started playing around with this. I _think_ it might make sense to put boundaries over all of our routes (maybe even build it into `AsyncRoute`?) and then if we have a particularly buggy component (ie beta feature) we could go deeper.

Reply:
It could be useful to frame this as a UX problem: In the case that something goes wrong, do we want to catch it higher up the render tree, away from the originating error context, or do we want to sprinkle these error boundaries throughout the tree at anticipated error locations such that the message visually identifies the source of the error?

Second, when an error occurs, should we be adding some sort of banner / message area above the affected tree, or should we just totally remove the tree as is done here?

My thought is that if we're catching errors higher up the tree (e.g. routes), then it would be a less jarring UX if we were to simply prepend the error message to the existing tree. But if we are to catch errors close to the originating locality, then we're probably dealing with visually smaller components, and can probably just replace their render tree with the message.